### PR TITLE
Update self.caller and self.signer documentation

### PR DESCRIPTION
**Pull request by craigjson**
Makes the following updates to the `.aleo` documentation
1. Updates existing references to `self.caller` to reference `self.signer`
2. Adds a section for `self.caller` explaining the new feature
3. Provides examples of the difference between `self.signer` and `self.caller`

One question, with this update, there will be inconsistency between the `Leo` and `Aleo` documentation. Namely, that because Leo has not yet implemented the new `self.caller`, the documentation that refers to it https://developer.aleo.org/leo/language#selfcaller will actually be talking about `self.signer` in Aleo instructions. I worry this may cause some confusion. We could remove this section in the Leo docs for now, curious to your thoughts @collinc97 

Part of: https://github.com/AleoHQ/snarkVM/pull/2020

### CODE REVIEW

File: documentation/aleo/03_language.md [hunk 1]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+Consider the following program as a motivating example for the usage of `self.signer` and `self.caller`
+
+```
+import import.aleo;
+
+program parent.aleo;
+
+record who_called_parent:
+    owner as address.private;
+    gates as u64.private;
```
Comment (collinc97) [inline documentation/aleo/03_language.md:453]:
```suggestion
```

File: documentation/aleo/03_language.md [hunk 2]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+```
+
+The program below, `import.aleo` is used in the above example to motivate the usage of `self.caller` and `self.signer`.
+
+```
+program import.aleo;
+
+record who_called_import:
+    owner as address.private;
+    gates as u64.private;
```
Comment (collinc97) [inline documentation/aleo/03_language.md:486]:
```suggestion
```

File: documentation/aleo/03_language.md [hunk 3]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+
+Consider the following program as a motivating example for the usage of `self.signer` and `self.caller`
+
+```
+import import.aleo;
+
+program parent.aleo;
+
+record who_called_parent:
+    owner as address.private;
```
Comment (collinc97) [inline documentation/aleo/03_language.md:452]:
```suggestion
    // The address of the owner.
    owner as address.private;
```

File: documentation/aleo/03_language.md [hunk 4]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+
+```
+
+The program below, `import.aleo` is used in the above example to motivate the usage of `self.caller` and `self.signer`.
+
+```
+program import.aleo;
+
+record who_called_import:
+    owner as address.private;
```
Comment (collinc97) [inline documentation/aleo/03_language.md:485]:
```suggestion
    // The address of the owner.
    owner as address.private;
```

File: documentation/aleo/03_language.md [hunk 5]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+
+The `self.caller` command returns the address of the immediate caller of the program.
+
+#### Usage of `self.signer` and `self.caller`
+
+Consider the following program as a motivating example for the usage of `self.signer` and `self.caller`
+
+```
+import import.aleo;
+
```
Comment (collinc97) [inline documentation/aleo/03_language.md:448]: Can you add a comment explaining the purpose of this program?

File: documentation/aleo/03_language.md [hunk 6]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+    // self.signer, should be original invoker of the program calling this one
+    caller as address.private;
+    // self.caller, should be the direct program address calling this one
+    parent as address.private;
+
+function external_call:
+    cast self.signer 0u64 self.signer self.caller into r0 as who_called_import.record;
+
+    output r0 as who_called_import.record;
+```
```
Comment (collinc97) [inline documentation/aleo/03_language.md:497]: Can you add some executable snarkvm CLI commands that will work when this program is copied and pasted into a project?

File: documentation/aleo/03_language.md [hunk 7]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+
+```
+import import.aleo;
+
+program parent.aleo;
+
+record who_called_parent:
+    owner as address.private;
+    gates as u64.private;
+    // self.signer, should be original invoker of the program calling this one
```
Comment (collinc97) [inline documentation/aleo/03_language.md:454]:
```suggestion
    // The address of the caller.
```
It's not clear to me why this "should be the original invoker of the program".
`caller` is a private address within a record data type.
I can import this record into another program and set `caller` to any arbitrary address.
`caller` cannot be restricted to a specific address at this point in the program.

File: documentation/aleo/03_language.md [hunk 8]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+    caller as address.private;
+    // self.caller, should be the direct program address calling this one
```
Comment (collinc97) [inline documentation/aleo/03_language.md:456]:
```suggestion
    // The address of the parent.
```
Please see my previous comment on line 454

File: documentation/aleo/03_language.md [hunk 9]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+    parent as address.private;
+
```
Comment (collinc97) [inline documentation/aleo/03_language.md:458]: Please add a comment describing what this function computes and returns.

File: documentation/aleo/03_language.md [hunk 10]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+The program below, `import.aleo` is used in the above example to motivate the usage of `self.caller` and `self.signer`.
+
+```
+program import.aleo;
+
+record who_called_import:
+    owner as address.private;
+    gates as u64.private;
+    // self.signer, should be original invoker of the program calling this one
```
Comment (collinc97) [inline documentation/aleo/03_language.md:487]:
```suggestion
    // The address of the caller.
```
Please see my previous comment on line 454

File: documentation/aleo/03_language.md [hunk 11]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+    caller as address.private;
+    // self.caller, should be the direct program address calling this one
+    parent as address.private;
```
Comment (collinc97) [inline documentation/aleo/03_language.md:490]:
```suggestion
    // The address of the parent.
```
Please see my previous comment on line 454

File: documentation/aleo/03_language.md [hunk 12]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+    output r2 as who_called_parent.record;
+    output r3 as import.aleo/who_called_import.record;
+
+```
+
+The program below, `import.aleo` is used in the above example to motivate the usage of `self.caller` and `self.signer`.
+
+```
+program import.aleo;
+
```
Comment (collinc97) [inline documentation/aleo/03_language.md:483]: Can you add a comment explaining the purpose of this program?

File: documentation/aleo/03_language.md [hunk 13]
```diff
@@ -429,10 +429,71 @@ finalize transfer_public_to_private:
...
+record who_called_import:
+    owner as address.private;
+    gates as u64.private;
+    // self.signer, should be original invoker of the program calling this one
+    caller as address.private;
+    // self.caller, should be the direct program address calling this one
+    parent as address.private;
+
```
Comment (collinc97) [inline documentation/aleo/03_language.md:491]: Please add a comment describing what this function computes and returns.

**Review by collinc97 — CHANGES_REQUESTED** [review]
Thanks for the PR @craigjson 

I think the motivating example is a good way to introduce this command.
Do you think developers will copy and paste this example into their projects?
Do you have additional snarkvm CLI commands that you ran to test the functionality?

**Comment by collinc97** [comment]
> Makes the following updates to the `.aleo` documentation
> 
> 1. Updates existing references to `self.caller` to reference `self.signer`
> 2. Adds a section for `self.caller` explaining the new feature
> 3. Provides examples of the difference between `self.signer` and `self.caller`
> 
> One question, with this update, there will be inconsistency between the `Leo` and `Aleo` documentation. Namely, that because Leo has not yet implemented the new `self.caller`, the documentation that refers to it https://developer.aleo.org/leo/language#selfcaller will actually be talking about `self.signer` in Aleo instructions. I worry this may cause some confusion. We could remove this section in the Leo docs for now, curious to your thoughts @collinc97
> 
> Part of: [AleoHQ/snarkVM#2020](https://github.com/AleoHQ/snarkVM/pull/2020)

I expect the feature parity between Leo and Aleo Instructions to be resolved this week.
For now you could add an [Admonition](https://docusaurus.io/docs/markdown-features/admonitions) that makes it clear for people that there is a difference.

**Comment by craigjson** [comment]
> Thanks for the PR @craigjson
> 
> I think the motivating example is a good way to introduce this command. Do you think developers will copy and paste this example into their projects? Do you have additional snarkvm CLI commands that you ran to test the functionality?

A developer can copy and paste this example, however I'd say it's fairly academic and limited in it's usability for a production program. It reminds me of an example a professor would use to show how OOP Inheritance works.

I added some steps to create a project and run this on the CLI to generate the output that is provided

**Comment by collinc97** [comment]
Hi @craigjson can you see my latest PR against this one https://github.com/AleoHQ/welcome/pull/291

**Comment by craigjson** [comment]
> Hi @craigjson can you see my latest PR against this one #291

merged

**Review by collinc97 — APPROVED** [review]
LGTM

**Pull request merged**